runtime/compiler/optimizer/OSRGuardInsertion.cpp

```diff
@@ -30,63 +30,205 @@
 #include "optimizer/Optimization_inlines.hpp"
 #include "optimizer/FearPointAnalysis.hpp"
 #include "optimizer/RematTools.hpp"
-#include "optimizer/HCRGuardAnalysis.hpp"
 #include "optimizer/TransformUtil.hpp"
 #include "ras/DebugCounter.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+
+TR_Structure* fakeRegion(TR::Compilation *comp);
 
 static bool generatesFear(TR::Compilation *comp, TR_FearPointAnalysis &fearAnalysis, TR::Block *block)
    {
    return fearAnalysis._blockAnalysisInfo[block->getNumber()]->get(0);
    }
 
-int32_t TR_OSRGuardInsertion::perform()
+
+static bool hasHCRGuard(TR::Compilation *comp)
    {
-   static char *disableOSRGuards = feGetEnv("TR_DisableOSRGuards");
+   TR::list<TR_VirtualGuard*> &virtualGuards = comp->getVirtualGuards();
+   auto guard = virtualGuards.begin();
+   for (; guard != virtualGuards.end(); ++guard)
+      {
+      if ((*guard)->getKind() == TR_HCRGuard || (*guard)->mergedWithHCRGuard())
+         return true;
+      }
 
-   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
-   if (disableOSRGuards || !comp()->isOSRTransitionTarget(TR::postExecutionOSR))
+   return true;
+   }
+
+static bool hasYieldPointCannotOSR(TR::Compilation *comp)
```

Comment:
I would prefer hasUnsupportedPotentialOSRPoint since this is about OSR points and the relationship to yield points is not determined - we are only concerned about transition points (which are normally yield points, but we don't need to conflate the two concepts).

Reply:
While your suggested name would be more consistent with the naming of the queries in terms of which this is implemented, I think it's those queries that are confusing. "OSR point" sounds like a point at which OSR is possible, a transition point as you put it. An example of the confusion this causes: this function is not concerned with transition points at all. It is concerned instead with points that allow assumptions to be invalidated but which are specifically _not_ transition points. In the context of OSR specifically, I think the "yield point" terminology does make it clear which property is relevant, although it does conflict somewhat with the more general notion of a yield point, and with the existing queries `isPotentialOSRPoint*`.

Maybe we should try to move things consistently to a set of clearer terminology? e.g. abandon "OSR point" and speak only of "OSR transition point" and "OSR yield point"/"OSR invalidation point." Not as part of this PR, of course.